cpp/include/cuvs/neighbors/quantization.hpp

```diff
@@ -0,0 +1,195 @@
+ */
+
+#pragma once
+
+#include <cuvs/neighbors/common.hpp>
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/handle.hpp>
+#include <raft/core/host_mdarray.hpp>
+#include <raft/core/host_mdspan.hpp>
+
+#include <cuda_fp16.h>
+
+namespace cuvs::neighbors::quantization {
+
+/**
+ * @brief ScalarQuantizer parameters.
+ */
+struct sq_params {
+  /*
+   * specifies how many outliers at top & bottom will be ignored
+   * needs to be within range of (0, 1]
+   */
+  float quantile = 0.99;
+};
+
+/**
+ * @brief Defines and stores scalar for quantisation upon training
+ *
+ * The quantization is performed by a linear mapping of an interval in the
+ * float data type to the full range of the quantized int type.
+ *
+ * @tparam T data element type
+ * @tparam QuantI quantized type of data after transform
+ *
+ */
+template <typename T, typename QuantI>
+class ScalarQuantizer {
+ public:
+  /**
+   * @brief Computes the scaling factor to be used later for quantizing the dataset.
+   *
+   * Usage example:
+   * @code{.cpp}
+   * raft::handle_t handle;
+   * cuvs::neighbors::quantization::ScalarQuantizer<float, int8_t> quantizer;
+   * cuvs::neighbors::quantization::sq_params params;
+   * quantizer.train(handle, params, dataset);
+   * @endcode
+   *
+   * @param[in] res raft resource
+   * @param[in] params configure scalar quantizer, e.g. quantile
+   * @param[in] dataset a row-major matrix view on device
+   */
+  void train(raft::resources const& res,
+             sq_params params,
+             raft::device_matrix_view<const T, int64_t> dataset);
+
+  /**
+   * @brief Computes the scaling factor to be used later for quantizing the dataset.
+   *
+   * Usage example:
+   * @code{.cpp}
+   * raft::handle_t handle;
+   * cuvs::neighbors::quantization::ScalarQuantizer<float, int8_t> quantizer;
+   * cuvs::neighbors::quantization::sq_params params;
+   * quantizer.train(handle, params, dataset);
+   * @endcode
+   *
+   * @param[in] res raft resource
+   * @param[in] params configure scalar quantizer, e.g. quantile
+   * @param[in] dataset a row-major matrix view on host
+   */
+  void train(raft::resources const& res,
+             sq_params params,
+             raft::host_matrix_view<const T, int64_t> dataset);
+
+  /**
+   * @brief Applies quantization transform to given dataset
+   *
+   * Requires train step to be finished.
+   *
+   * Usage example:
+   * @code{.cpp}
+   * raft::handle_t handle;
+   * cuvs::neighbors::quantization::ScalarQuantizer<float, int8_t> quantizer;
+   * cuvs::neighbors::quantization::sq_params params;
+   * quantizer.train(handle, params, dataset);
+   * auto quantized_dataset = quantizer.transform(handle, dataset);
+   * @endcode
+   *
+   * @param[in] res raft resource
+   * @param[in] dataset a row-major matrix view on device
+   *
+   * @return device matrix with quantized dataset
+   */
+  raft::device_matrix<QuantI, int64_t> transform(
```

Comment:
Most of the functions in cuvs usually take mdspan as inputs/outputs, but here you are using device arrays as return values. Is it possible to use mdspan as in/out to be as consistent as possible with the other functions of the cuvs API?